src/query_plan/operation.rs

```diff
@@ -60,13 +62,113 @@ pub(crate) struct NormalizedSelectionSet {
     pub(crate) selections: Arc<NormalizedSelectionMap>,
 }
 
-/// A "normalized" selection map is an optimized representation of a selection set which does not
-/// contain selections with the same selection "key". Selections that do have the same key are
-/// merged during the normalization process. By storing a selection set as a map, we can efficiently
-/// merge/join multiple selection sets.
-///
-/// Note that this must be a `LinkedHashMap` so that removals don't change the order.
-pub(crate) type NormalizedSelectionMap = LinkedHashMap<NormalizedSelectionKey, NormalizedSelection>;
+pub(crate) mod normalized_selection_map {
+    use crate::query_plan::operation::{NormalizedSelection, NormalizedSelectionKey};
+    use linked_hash_map::LinkedHashMap;
+    use std::borrow::Borrow;
+    use std::hash::Hash;
+    use std::ops::Deref;
+
+    /// A "normalized" selection map is an optimized representation of a selection set which does
+    /// not contain selections with the same selection "key". Selections that do have the same key
+    /// are  merged during the normalization process. By storing a selection set as a map, we can
+    /// efficiently merge/join multiple selection sets.
+    ///
+    /// Because the key depends strictly on the value, we expose the underlying map's API in a
+    /// read-only capacity, while mutations use an API closer to `IndexSet`. We don't just use an
+    /// `IndexSet` since key computation is expensive (it involves sorting). This type is in its own
+    /// module to prevent code from accidentally mutating the underlying map outside the mutation
+    /// API.
+    ///
+    /// Note that this must be a `LinkedHashMap` so that removals don't change the order.
```

Comment:
We can also preserve order in `IndexMap` by using `shift_remove` instead of `swap_remove`.  I’ve filed https://github.com/apollographql/federation-next/issues/140

Reply:
This was easy enough, so I made the change in https://github.com/apollographql/federation-next/pull/135/commits/7daa28f06618b0b087b563cedc01ce3d9e247ba1